topomodelx/nn/combinatorial/hmc_layer.py

```diff
@@ -0,0 +1,1075 @@
+            ).squeeze(1),
+            size=(s_message.shape[0], t_message.shape[0]),
+            device=self.get_device(),
+        )
+
+        if self.softmax:
+            return torch.sparse.softmax(e, dim=1), torch.sparse.softmax(f, dim=1)
+
+        return sparse_row_norm(e), sparse_row_norm(f)
+
+    def forward(
+        self, x_source: torch.Tensor, x_target: torch.Tensor, neighborhood: torch.Tensor
+    ) -> tuple[torch.Tensor]:
+        r"""Compute forward pass.
+
+        The forward pass of the Higher Order Attention Block for non-squared
+        matrices (HBNS) is defined as
+
+        ..  math::
+            \begin{align}
+                \text{HBNS}_N(X_s, X_t) = (Y_s, Y_t),
+            \end{align}
+
+        where the source and target outputs :math:`Y_s` and :math:`Y_t` are
+        computed as
+
+        ..  math::
+            \begin{align}
+                Y_s &= \phi((N^T \odot A_t) X_t W_t), \\
+                Y_t &= \phi((N \odot A_s) X_s W_s ).
+            \end{align}
+
+        Parameters
+        ----------
+        x_source : torch.Tensor, shape=[source_cells, source_in_channels]
+            Cochain matrix representation :math:`X_s` containing the signal
+            features over the source cells.
+        x_target : torch.Tensor, shape=[target_cells, target_in_channels]
+            Cochain matrix :math:`X_t` containing the signal features over
+            the target cells.
+        neighborhood : torch.sparse, shape=[target_cells, source_cells]
+            Neighborhood matrix :math:`N` inducing the HBNS block.
+
+        Returns
+        -------
+        _ :math:`Y_s` : torch.Tensor, shape=[source_cells, source_out_channels]
+            Output features of the layer for the source cells.
+        _ :math:`Y_t` : torch.Tensor, shape=[target_cells, target_out_channels]
+            Output features of the layer for the target cells.
+        """
+        s_message = torch.mm(x_source, self.w_s)  # [n_source_cells, d_t_out]
+        t_message = torch.mm(x_target, self.w_t)  # [n_target_cells, d_s_out]
+
+        neighborhood_s_to_t = (
+            neighborhood.coalesce()
+        )  # [n_target_cells, n_source_cells]
+        neighborhood_t_to_s = (
+            neighborhood.t().coalesce()
+        )  # [n_source_cells, n_target_cells]
+
+        self.target_indices, self.source_indices = neighborhood_s_to_t.indices()
+
+        s_to_t_attention, t_to_s_attention = self.attention(s_message, t_message)
+
+        neighborhood_s_to_t_att = torch.sparse_coo_tensor(
+            indices=neighborhood_s_to_t.indices(),
+            values=s_to_t_attention.values() * neighborhood_s_to_t.values(),
+            size=neighborhood_s_to_t.shape,
+            device=self.get_device(),
+        )
+
+        neighborhood_t_to_s_att = torch.sparse_coo_tensor(
+            indices=neighborhood_t_to_s.indices(),
+            values=t_to_s_attention.values() * neighborhood_t_to_s.values(),
```

Comment:
According to the types, `t_to_s_attention` and `neighbourhood_t_to_s` are tensors, which do not have a `values` function.

Reply:
Totally right! Suggested to change types to torch.sparse.FloatTensor in attention method.